### PR TITLE
[css-fonts-4] Contribute Microsoft tests for Font Variations

### PR DESCRIPTION
**Pull request by FremyCompany**
Contributing (most) of our own CSS Fonts 4 tests to web-platform-tests. Some tests have not been ported because they require Microsoft Windows fonts and I have not enough time/expertise to create a dummy version of these fonts.

**INTEROP:** Chrome, Safari and Firefox all seem to have issues with these tests, so it would be good if someone from either/all of these browsers could review the tests. 

**NOTE:** One of the test requires you to install some fonts natively on your system. These fonts are a new (but backward-compatible) version of the fonts that are already required for other css-fonts tests. I created this new version because the old version didn't give any mean to test font matching via JavaScript, which is annoying. If this PR gets merged, someone should update http://www.w3.org/Style/CSS/Test/Fonts/CSSTest/ but I don't know who to contact.

### CODE REVIEW

**Comment by w3c-bots** [comment]
# Build PENDING

Started: None
Finished: None


<h2>Failing Jobs</h2>
<ul>

<li>chrome:dev</li>

</ul>




View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/23822)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/336770565)

File: css/css-fonts/variations/font-opentype-collections.html
Comment (SergeyMalkin) [inline css/css-fonts/variations/font-opentype-collections.html]: This is not variable fonts test, so you may want to move it outside variations folder. But otherwise test is useful and should also be expanded to test access to individual fonts inside collection.

File: css/css-fonts/variations/font-shorthand.html
Comment (SergeyMalkin) [inline css/css-fonts/variations/font-shorthand.html]: Comment is incorrect. This should be something like "font-weight is not part of shorthand"

File: css/css-fonts/variations/font-style.html
Comment (SergeyMalkin) [inline css/css-fonts/variations/font-style.html]: This test is already in separate file. You can remove it from here.

Comment (SergeyMalkin) [reply css/css-fonts/variations/font-style.html]: Same for transition test below

File: css/css-fonts/variations/font-weight-interpolation.html
Comment (SergeyMalkin) [inline css/css-fonts/variations/font-weight-interpolation.html]: fontStyleAnimation -> fontWeightAnimation

File: css/css-fonts/variations/font-variation-settings-parsing.html
Comment (SergeyMalkin) [inline css/css-fonts/variations/font-variation-settings-parsing.html]: Add "use strict"

File: css/css-fonts/variations/font-variation-settings-parsing.html
Comment (SergeyMalkin) [inline css/css-fonts/variations/font-variation-settings-parsing.html]: Would be nice to have names consistent between tests, e.g. input/output/valid->value/expectedComputedValue/isValid

File: css/css-fonts/variations/font-weight-lighter-bolder.html
Comment (SergeyMalkin) [inline css/css-fonts/variations/font-weight-lighter-bolder.html]: "use strict"

File: css/css-fonts/variations/font-weight-parsing.html
Comment (SergeyMalkin) [inline css/css-fonts/variations/font-weight-parsing.html]: expectedStyle -> expectedWeight

**Comment by FremyCompany** [comment]
Linting errors I will have to fix:
- TRAILING WHITESPACE: 1 
- AHEM COPY: 1 
- CR AT EOL: 940 
- MISSING-LINK: 13

**Comment by SergeyMalkin** [comment]
You forgot to port two files: extensive set of font matching tests and also tests for parsing @font-face descriptors.

**Review by SergeyMalkin — CHANGES_REQUESTED** [review]

**Comment by emilio** [comment]
cc @jfkthame

File: css/css-fonts/variations/font-style-parsing.html
Comment (SergeyMalkin) [inline css/css-fonts/variations/font-style-parsing.html]: Can you add cases for different angle units? E.g. calc(2 * 0.1rad)

Comment (FremyCompany) [reply css/css-fonts/variations/font-style-parsing.html]: I doubt that this is relevant. A browser that would fail these tests but not the existing ones would have an issue with css-value-and-units not font-style parsing.

Comment (SergeyMalkin) [inline css/css-fonts/variations/font-style-parsing.html]: Are you calc(-120deg) is valid value? My understanding is that final calculated value is tested against acceptable range, otherwise it is treated as parse error.

Comment (FremyCompany) [reply css/css-fonts/variations/font-style-parsing.html]: https://drafts.csswg.org/css-values/#calc-range

Comment (SergeyMalkin) [reply css/css-fonts/variations/font-style-parsing.html]: Ah, ok. Then both Edge and Chrome have the same bug across font-weight/stretch/style.

Can you review font-weight and font-stretch as well?

**Comment by drott** [comment]
> INTEROP: Chrome, Safari and Firefox all seem to have issues with these tests, so it would be good if someone from either/all of these browsers could review the tests.

Thanks for upstreaming and contributing these, @FremyCompany. I will put it on my TODO list to go through them for Blink and leave review feedback accordingly, but it will take a bit more time.

File: css/css-fonts/variations/font-style-parsing.html
Comment (SergeyMalkin) [reply css/css-fonts/variations/font-style-parsing.html]: Are there tests for css-values to cover calc() for angles? I don't see any, but I may be missing it. 

There are specific tests covering other properties and units. We should have tests for angles, either here or in css-values folder

**Comment by FremyCompany** [comment]
@drott Thanks, totally understood

**Review by drott — CHANGES_REQUESTED** [review]
An excellent quality set of test cases - very helpful, thank you! Running these in Blink already uncovered 9 issues (compare [this meta bug](https://bugs.chromium.org/p/chromium/issues/detail?id=809920)).

I got curious about these test cases so I gave them a go in Blink. 2 pass as is, 10 fail, some of them only with minor issus.

For the test cases that partially need system fonts, I would suggest to split those out into separate tests for making it easier to compare results across engines, as the manual installation / harness preparation with the system fonts is an extra step that automated testing might not cover well yet.

I don't agree with the expectations for `css/css-fonts/variations/font-variation-settings-inherit.html`. But this might be only my interpretation of the spec or I might have missed something. It seems we need a spec clarification here. If you'd like to merge this PR quickly, I would suggest to isolate this one into a separate PR until we work out with @litherum   and others what the expected behavior should be.

Some test cases trigger a harness error / warning for duplicate test names. I believe this is minor but probably worth fixing.

Thanks again, I am happy to straighten out the Blink issues to pass more of these new tests.

File: css/css-fonts/variations/at-font-face-font-matching.html
Comment (FremyCompany) [reply css/css-fonts/variations/at-font-face-font-matching.html]: Seems reasonable, will do

File: css/css-fonts/variations/font-shorthand.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-shorthand.html]: Yes, I can make the calc land on 0.5 multiple.

File: css/css-fonts/variations/font-shorthand.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-shorthand.html]: On second look though, this expression should be valid though, because of calc() clamping. 
I will update the expectation here ^_^

File: css/css-fonts/variations/font-shorthand.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-shorthand.html]: I think it's fine to leave the test strict here, but it's also probably ok not to pass that specific test, I doubt this is going to cause much webcompat trouble ^_^

I think there's still supposed to be a different between oblique and italic if a font supports both. It seems `oblique` shoud (I *think* but I'm not entirely sure) prefer any oblique fontface over the italic fontface, while italic should prefer any italic version of an oblique fontface.

I have never seen a font with both an oblique and an italic variant though, so I am not sure how to test whether this is true in any browser. Maybe the LaTeX font?

File: css/css-fonts/variations/font-variation-settings-inherit.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-variation-settings-inherit.html]: It might be a result of how we store the flags in Edge, maybe we just can't store this duplication internally and therefore simplify anyway. Paging @SergeyMalkin here.

File: css/css-fonts/variations/font-weight-parsing.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-weight-parsing.html]: I will change to 400.5

File: css/css-fonts/variations/font-weight-matching.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-weight-matching.html]: Seems reasonable

File: css/css-fonts/variations/font-variation-settings-inherit.html
Comment (csnardi) [reply css/css-fonts/variations/font-variation-settings-inherit.html]: Drive-by comment: I believe the spec should soon be updated to reflect this as correct behavior per https://github.com/w3c/csswg-drafts/issues/1959.

File: css/css-fonts/variations/font-shorthand.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-shorthand.html]: ✔

File: css/css-fonts/variations/font-shorthand.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-shorthand.html]: ✔

File: css/css-fonts/variations/font-shorthand.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-shorthand.html]: ✔

File: css/css-fonts/variations/font-shorthand.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-shorthand.html]: ✔ (no change)

File: css/css-fonts/variations/font-shorthand.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-shorthand.html]: ✔ (no change)

File: css/css-fonts/variations/font-stretch.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-stretch.html]: ✔ (the same test was just duplicated)

File: css/css-fonts/variations/font-style-parsing.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-style-parsing.html]: ✔ (no change)

File: css/css-fonts/variations/font-variation-settings-inherit.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-variation-settings-inherit.html]: ✔ (no change)

File: css/css-fonts/variations/font-style-parsing.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-style-parsing.html]: ✔ (no change)

File: css/css-fonts/variations/font-variation-settings-parsing.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-variation-settings-parsing.html]: ✔

File: css/css-fonts/variations/font-shorthand.html
Comment (jfkthame) [reply css/css-fonts/variations/font-shorthand.html]: Yes, the Computer Modern (or the newer Latin Modern) family offers both oblique ("slanted") and italic, and indeed they are used contrastively by some authors (e.g. Knuth himself).

(It also has an Unslanted Italic face, which AFAICS is not something we could express with CSS according to the current draft. I suppose something like

    font-style: italic oblique(0deg);

would make sense, if we want to allow it, but currently that would be a parse error.)

CSS Fonts 3 made it quite clear that 'italic' and 'oblique' could be distinguished, at least in `@font-face` families even if the UA doesn't support a distinction for system font families. I don't see any reason why we should regress that in Fonts 4.

File: css/css-fonts/variations/font-weight-parsing.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-weight-parsing.html]: ✔

File: css/css-fonts/variations/at-font-face-font-matching.html
Comment (FremyCompany) [reply css/css-fonts/variations/at-font-face-font-matching.html]: ✔ (this test didn't require the fonts to be installed, I converted it to use linked fonts)

**Comment by FremyCompany** [comment]
> For the test cases that partially need system fonts, I would suggest to split those out into separate tests for making it easier to compare results across engines, as the manual installation / harness preparation with the system fonts is an extra step that automated testing might not cover well yet.

Done. 

I also converted to url() fonts some tests that unnecessarily relied on the fonts being available. 

> I don't agree with the expectations for css/css-fonts/variations/font-variation-settings-inherit.html. But this might be only my interpretation of the spec or I might have missed something. It seems we need a spec clarification here. If you'd like to merge this PR quickly, I would suggest to isolate this one into a separate PR until we work out with @litherum and others what the expected behavior should be.

I updated the test to have a generic-enough assertion it passes in both browsers. We can always refine in a further pull request when we have the expected behavior clarified in the spec.

**Review by drott — APPROVED** [review]
Thanks for addressing the comments, LGTM. I'll file an issue for the inheritance / precedence of duplicate axis tags questions.

File: css/css-fonts/variations/font-variation-settings-inherit.html
Comment (drott) [inline css/css-fonts/variations/font-variation-settings-inherit.html]: Filed https://github.com/w3c/csswg-drafts/issues/2288

Comment (svgeesus) [reply css/css-fonts/variations/font-variation-settings-inherit.html]: @drott did you want to hold off on approving all tests while that issue gets discussed, or approve all but that test, or approve all and then fix up if needed?

File: css/css-fonts/variations/font-shorthand.html
Comment (svgeesus) [reply css/css-fonts/variations/font-shorthand.html]: Indeed, Fonts 4 makes it clear that `oblique` takes an optional second parameter, the oblique angle; and  `oblique` now expands to `oblique 20deg`.

**Comment by svgeesus** [comment]
@FremyCompany 
> If this PR gets merged, someone should update http://www.w3.org/Style/CSS/Test/Fonts/CSSTest/ but I don't know who to contact.

That would be me. But discuss with CSS WG first as there have been several updates to Ahem over recent years and we want to be sure the new one is a true superset of the most recent previous update (plus there was quite some discussion as I recall).

File: css/css-fonts/variations/font-variation-settings-inherit.html
Comment (drott) [reply css/css-fonts/variations/font-variation-settings-inherit.html]: I am good with merging the more lenient version of this test and discussing on the separate issue in the meantime. 

I did approve the PR, see above "drott approved these changes 23 hours ago", but it seems like that doesn't count 😢.

"Review required
At least one approved review is required by reviewers with write access. "

Would I be eligible for write access, who could configure it for me?

**Comment by FremyCompany** [comment]
@svgeesus I didn't update Ahem so that shouldn't be a problem. 

I updated [the CSSTest fonts that are currently provided at the url pointed above](http://www.w3.org/Style/CSS/Test/Fonts/CSSTest/). I manually checked all the tests that are using those fonts, and they didn't regress based on the update, so I think we should be fine.

**Review by svgeesus — CHANGES_REQUESTED** [review]
Approved (once the build passes)

File: css/css-fonts/variations/font-variation-settings-inherit.html
Comment (svgeesus) [reply css/css-fonts/variations/font-variation-settings-inherit.html]: > Would I be eligible for write access, who could configure it for me?

I just did, https://github.com/w3c/web-platform-tests/pull/9525 but now that tiny edit needs a reviewer ...

**Comment by FremyCompany** [comment]
@svgeesus The build now passes everything but the Firefox run. It does not fail because of me, it fails because it cannot seem to start the test. After 24mins, it gave up. I think we should merge this, and someone from the infra team should take a look at what is going on.

I can't merge because your review is marked as "requiring changes" though.

**Comment by drott** [comment]
@svgeesus ping? Can you help getting this merged, thank you!

File: css/css-fonts/variations/font-variation-settings-parsing.html
Comment (csnardi) [reply css/css-fonts/variations/font-variation-settings-parsing.html]: This still seems to exist, looking at the Travis log:
```
`Property value: Percentages should not be supported` | **FAIL: 20/10** 
`@supports: Percentages should not be supported` | **PASS: 20/10**
```

File: css/css-fonts/variations/font-style-parsing.html
Comment (csnardi) [inline css/css-fonts/variations/font-style-parsing.html]: Another duplicate test name:

```
`Font-style: 'oblique' followed by positive out of range angle is in invalid` | **PASS: 20/10**
```

**Review by csnardi — COMMENTED** [review]
It looks like the Firefox build was actually failing, which would prevent this from being merged.

File: css/css-fonts/variations/font-style-parsing.html
Comment (FremyCompany) [reply css/css-fonts/variations/font-style-parsing.html]: Ah, that's what 20/10 means. That wasn't clear to me. Hopefully fixed now?

**Comment by FremyCompany** [comment]
Thanks @csnardi that was indeed it 🙌
@svgeesus would you mind merging the pull request now?

**Comment by FremyCompany** [comment]
Up @svgeesus

**Review by svgeesus — APPROVED** [review]

**Pull request merged**
